symfit/core/fit_results.py

```diff
@@ -69,10 +69,10 @@ def __getattr__(self, item):
         :param item: Name of Goodness of Fit qualifier.
         :return: Goodness of Fit qualifier if present.
         """
-        if item in self.gof_qualifiers:
-            return self.gof_qualifiers[item]
-        else:
-            raise AttributeError
+        if 'gof_qualifiers' in vars(self):
```

Comment:
Why not just do:
`return self.gof_qualifiers[item]`
?
If the attribute is not there, AttributeError is raised anyway.

Reply:
Upon pickle.load, gof_qualifiers does not yet exist and then this will cause an infinite recursion error. This is apparently the correct way to bypass it, see e.g. https://stackoverflow.com/questions/49380224/how-to-make-classes-with-getattr-pickable

Reply:
The first suggestion will not work and give the same recursion error. The problem is with `got_qualifiers` not existing, not with the key not existing. You are right, if an attribute doesn't exist, `__getattr__` is called. Then, if that raises an `AttributeError` too, that's the end of it.

However, calling `self. gof_qualifiers` if `self. gof_qualifiers` does not exist, raises an AttributeError, which takes you to `__getattr__`. Then, in `__getattr__`, we look for `self. gof_qualifiers`. The result is infinite recursion.

`__getinitargs__` is only for old-style classes, which we don't use. It seems the py3 equivalent is `__reduce__`. Although that might also be worth adding so that `__init__` is called, I think this fix of `__getattr__` is also in order.

Reply:
I made a small testclass, and it seems you're right :'( (and it's slowly starting to make sense).
Creating a simple `__getstate__` and `__setstate__` also fixes it though. This at least keeps `pickle` specifics in the place you expect them.

```python
class A:
    def __init__(self):
        self.attr = {}
        self.attr['a'] = 1

    def __getattr__(self, item):
        try:
            return self.attr[item]
        except KeyError as error:
            raise AttributeError from error

    def __getstate__(self):
        return self.__dict__.copy()

    def __setstate__(self, state):
        self.__dict__.update(state)


a = A()

p = pickle.dumps(a)

a2 = pickle.loads(p)
```

Reply:
I agree that maybe that should be done too, but updating `__getattr__` in this way does make sure this function is at least consistent, do you agree?

I think it is weird that such minimal getstate and setstate change anything btw, because this is kinda what I expect python to do anyway.

Reply:
Implementing `__reduce__` on all models such that init is called upon pickle.load seems to solve all the problems. If you agree with this solution then I will merge this PR so our mystery power-user @Jhsmit will be happy.

I will then also make a new release, 0.4.5.1, to trigger a new Zenodo DOI build to see if that has been resolved.